kopf/clients/auth.py

```diff
@@ -1,42 +1,125 @@
 import logging
+from typing import Optional
 
-import kubernetes
+import pykube
+import requests
 import urllib3.exceptions
 
 logger = logging.getLogger(__name__)
 
+# Set in login(), consumed in get_pykube_cfg() and all API calls.
+_pykube_cfg: Optional[pykube.KubeConfig] = None
+
 
 class LoginError(Exception):
     """ Raised when the operator cannot login to the API. """
 
 
-def login():
+class AccessError(Exception):
+    """ Raised when the operator cannot access the cluster API. """
+
+
+def login(verify=False):
     """
-    Login the the Kubernetes cluster, locally or remotely.
+    Login to Kubernetes cluster, locally or remotely.
+
+    Keep the logged in state or config object in the global variables,
+    so that it can be available for future calls via the same function call.
+
+    Automatic refresh/reload of the tokens or objects also should be done here.
     """
 
-    # Configure the default client credentials for all possible environments.
+    # Pykube login is mandatory. If it fails, the framework will not run at all.
+    try:
+        import pykube
+    except ImportError:
+        raise  # mandatory
```

Comment:
`pykube` is now imported at the module-level, I suppose this can be removed.

Reply:
I thought about the future when pykube-ng could also be replaced with something, e.g. with the raw `requests-async` — for the sake of async HTTP calls. Total asynchronicity is one of the mid-term goals (not a priority at the moment though, as it is "solved" with the thread-pools). In that case, these libraries can be used to parse the config/envvars only, to retrieve the token(s). Or not used at all.